// yh-plugin-project/webpack.config.js
var HappyPack = require('happypack');

let exportsMap = ["Funnel.js","Pie.js"];

module.exports = exportsMap.map((name) => ({
   devtool: 'cheap-module-inline-source-map',
   entry: {
      [name]: './plugin/' + name
   },
   output: {
      path:  __dirname + "/plugin/result",
      filename: name,
   },
   optimization: {
      splitChunks: false
   },
   plugins: [
      new HappyPack({
         id: 'jsx',
         threads: 2,
         loaders: [
            {
               loader: "babel-loader",
               exclude: /node_modules/,
               options: {
                  presets: ['react', 'es2015', 'stage-0'],
                  cacheDirectory: true,
               }
            },
            {
               loader: "eslint-loader",
               enforce: 'pre',
               exclude: /node_modules/,
               options: {
                  configFile: './.eslintrc.js'
               }
            }
         ],
      }),
      new HappyPack({
         id: 'scss',
         threads: 2,
         loaders: ["style-loader", "css-loader", "postcss-loader", "sass-loader"],
      }),
      new HappyPack({
         id: 'css',
         threads: 2,
         loaders: ["style-loader", "css-loader"],
      }),
   ],
   resolve: {
      mainFields: ['jsnext:main', 'main'],
      alias: {
         moment: 'moment/moment.js'
      },
      extensions: ['.js', '.jsx'],
      symlinks: false
   },
   module: {
      rules: [
         {
            test: /\.jsx?$/,
            exclude: /node_modules/,
            use: ['happypack/loader?id=jsx'],
         },
         {
            test: /\.css$/, // Only .css files
            use: ['happypack/loader?id=css'] // Run both loaders
         },
         {
            test: /\.scss$/,
            use: ['happypack/loader?id=scss']
         },
         {
            test: /\.(ttf|eot|svg|woff(2)?)(\?[a-z0-9]+)?$/,
            use: [{
               loader: 'url-loader',
               options: {
                  limit: 100000
               }
            }]
         }
      ],
   },
}));
